### PR TITLE
Add ManagedDependencies folder to Function App module path.

### PR DESCRIPTION
**Pull request by Francisco-Gamino**
Add ManagedDependencies folder to Function App module path:

* When a new Function App is initialized locally (via func init), the managedDependencies setting will be added to the host.json on the local machine for PowerShell. The initialization of the Function App will download the latest Azure (Az) modules and place them in the ManagedDependencies folder in the Function App. When a function is uploaded, the values in host.json is read and this is applied to the Function App. The local modules  in the ManagedDependencies folder are not packaged up and uploaded to the service.

* During function runtime in the service, if the managedDependencies setting is set, the latest module version of the modules are used in the service. Azure Functions has a service in the back end which manage minor updates for the Azure modules and makes them available to the customer at runtime. 

* Module’s folder structure for a Function App:
c:\MyFunction\Modules                            --> These are the customer modules
c:\MyFunction\ManagedDependencies    --> These are modules managed by Azure Functions

### Client UX from CLI + PowerShell Core
```powershell
<#
Prerequisites: 
    - PowerShell Core
    - Func.exe via CLI
#>

# Start PowerShell Core.
# Make sure the Functions CLI is part of the PATH environment. 

# Create a local directory where the function app will be created
New-Item -Path c:\MyFunction -ItemType Directory -Force

# Create a function app via func.exe. When running this command,
# the user gets prompt to install Managed Dependencies
func init MyFunctionApplication --worker-runtime powershell

# Prompt  the user
Would you like to install the Azure modules and have these automatically managed in Azure?
Yes

<#
# After this, the Az modules are installed under the Function App
c:\MyFunction\Modules
c:\MyFunction\ManagedDependencies\Az
c:\MyFunction\ManagedDependencies\Az\1.0.0
c:\MyFunction\ManagedDependencies\Az.Aks\1.0.0
c:\MyFunction\ManagedDependencies\Az.AnalysisServices\1.0.0
c:\MyFunction\ManagedDependencies\Az.ApiManagement\1.0.0
c:\MyFunction\ManagedDependencies\Az.ApplicationInsights\1.0.0
c:\MyFunction\ManagedDependencies\Az.Automation\1.0.0
c:\MyFunction\ManagedDependencies\Az.Batch\1.0.0
c:\MyFunction\ManagedDependencies\Az.Billing\1.0.0
c:\MyFunction\ManagedDependencies\Az.Cdn\1.0.0
# More Az modules...
#>

# Entry in Host.json for managed dependencies
Host.json file:
    "managedDependencies": [
        {
            "Name": "Az",
            "MajorVersion": "~1"            
        }
    ]

```

### CODE REVIEW

File: src/FunctionLoader.cs [hunk 1]
```diff
@@ -53,10 +53,18 @@ internal static void SetupWellKnownPaths(FunctionLoadRequest request)
...
             // Resolve the FunctionApp root path
             FunctionAppRootPath = Path.GetFullPath(Path.Join(request.Metadata.Directory, ".."));
+
             // Resolve module paths
             var appLevelModulesPath = Path.Join(FunctionAppRootPath, "Modules");
             var workerLevelModulesPath = Path.Join(AppDomain.CurrentDomain.BaseDirectory, "Modules");
-            FunctionModulePath = $"{appLevelModulesPath}{Path.PathSeparator}{workerLevelModulesPath}";
+
+            // Managed dependencies are modules managed by Azure Functions
+            var workerLevelManagedDependencies = Path.Join(AppDomain.CurrentDomain.BaseDirectory, "ManagedDependencies");
```
Comment (TylerLeonhardt) [inline src/FunctionLoader.cs]: > c:\MyFunction\ManagedDependencies --> These are modules managed by Azure Functions

You mention this in the description but your code actually puts this folder next to the PowerShell worker's dlls - NOT inside the Function App.

What is the expected behavior here?

**Comment by TylerLeonhardt** [comment]
So a couple things

---

> c:\MyFunction\MyHTTPTrigger\Modules --> These are the customer modules available only for the MyHTTPTrigger function

Just wanted to call this out that this does not exist. We figured users didn't need that yet. Willing to assess something like that in if we get feedback from users. Today, only the `c:\MyFunction\Modules` and the worker level Module folder is present.

---

What will be populating this folder and how will it be kept up-to-date?

---

What is the local experience like for ManagedDependencies?

---

**Comment by Francisco-Gamino** [comment]
@TylerLeonhardt : Thank you for comments. I have updated the PR description with additional information as well as the end to end experience.

File: src/FunctionLoader.cs [hunk 2]
```diff
@@ -53,10 +53,18 @@ internal static void SetupWellKnownPaths(FunctionLoadRequest request)
...
             // Resolve the FunctionApp root path
             FunctionAppRootPath = Path.GetFullPath(Path.Join(request.Metadata.Directory, ".."));
+
             // Resolve module paths
             var appLevelModulesPath = Path.Join(FunctionAppRootPath, "Modules");
             var workerLevelModulesPath = Path.Join(AppDomain.CurrentDomain.BaseDirectory, "Modules");
-            FunctionModulePath = $"{appLevelModulesPath}{Path.PathSeparator}{workerLevelModulesPath}";
+
+            // Managed dependencies are modules managed by Azure Functions
+            var appLevelManagedDependenciesPath = Path.Join(FunctionAppRootPath, "ManagedDependencies");
```
Comment (daxian-dbw) [inline src/FunctionLoader.cs]: The name **appLevel**ManagedDependenciesPath suggesting there is other level managed dependencies path. I recommend to just use `managedDependenciesPath` as the name here.

File: src/FunctionLoader.cs [hunk 3]
```diff
@@ -53,10 +53,18 @@ internal static void SetupWellKnownPaths(FunctionLoadRequest request)
...
+
+            // Set FunctionApp module path
+            FunctionModulePath = $"{appLevelModulesPath}{Path.PathSeparator}" +
+                                 $"{workerLevelModulesPath}{Path.PathSeparator}" +
+                                 $"{appLevelManagedDependenciesPath}";
```
Comment (daxian-dbw) [inline src/FunctionLoader.cs]: The last one don't need to be warped in a expandable string. Just use
```
$"{workerLevelModulesPath}{Path.PathSeparator}" +
appLevelManagedDependenciesPath
```

File: test/Unit/PowerShell/PowerShellManagerTests.cs [hunk 4]
```diff
@@ -148,7 +148,10 @@ public void ModulePathShouldBeSetCorrectly()
         {
             string workerModulePath = Path.Join(AppDomain.CurrentDomain.BaseDirectory, "Modules");
             string funcAppModulePath = Path.Join(FunctionLoader.FunctionAppRootPath, "Modules");
-            string expectedPath = $"{funcAppModulePath}{Path.PathSeparator}{workerModulePath}";
+            string funcAppManagedDependenciesPath = Path.Join(FunctionLoader.FunctionAppRootPath, "ManagedDependencies");
+            string expectedPath = $"{funcAppModulePath}{Path.PathSeparator}" +
+                                  $"{workerModulePath}{Path.PathSeparator}" +
+                                  $"{funcAppManagedDependenciesPath}";
```
Comment (daxian-dbw) [inline test/Unit/PowerShell/PowerShellManagerTests.cs]: same here.

**Comment by daxian-dbw** [comment]
@Francisco-Gamino According to the PR description, the `ManagedDependecies` folder in the local Function App folder won't get updated after the initialization, right? So when the user is updating the function locally and testing, he/she may be using a different version of the managed modules compared with when the function actually running in Azure ... how is that problem handled?

File: src/FunctionLoader.cs (hunk 2 above)
Comment (Francisco-Gamino) [reply src/FunctionLoader.cs]: Good point @daxian-dbw. var name updated.

**Comment by Francisco-Gamino** [comment]
This change is already included in https://github.com/Azure/azure-functions-powershell-worker/pull/124. Closing PR.

**Pull request closed without merging**